chef/cookbooks/cpe_activedirectory/resources/cpe_activedirectory.rb

```diff
@@ -8,70 +8,213 @@
 #
 
-require 'mixlib/shellout'
-
 resource_name :cpe_activedirectory
+provides :cpe_activedirectory
 default_action :run
 property :scope, default: 'all'
 
-def already_set?(root, key, value)
-  flag_to_profile_map = node['cpe_activedirectory']['flag_to_profile_map']
-  profile_root = root.split('_').map(&:capitalize).join(' ')
-  profile_key = flag_to_profile_map[root][key]
-  return nil if profile_key.nil?
+action :run do
+  unbind if remediate? || unbind?
+  bind if bind?
+  configure if configure?
+end
 
-  return nil if node.dsconfigad_profile.nil?
+action_class do # rubocop:disable Metrics/BlockLength
+  def bind?
+    node['cpe_activedirectory']['bind']
+  end
 
-  root_hash = node.dsconfigad_profile[profile_root]
-  return nil if root_hash.nil?
+  def configure?
+    node['cpe_activedirectory']['configure']
+  end
 
-  root_hash[profile_key] == value
-end
+  def remediate?
+    node['cpe_activedirectory']['remediate']
+  end
 
-def shell_format(value)
-  if value.is_a? Array
-    "'#{value.join(',')}'"
-  elsif [true, false].include? value
-    value ? 'enable' : 'disable'
-  else
-    value.to_s
+  def unbind?
+    node['cpe_activedirectory']['unbind']
   end
-end
 
-action :run do
-  return unless node['cpe_activedirectory']['configure']
+  def already_set?(ad_state, root, key, value)
+    flag_to_profile_map = node['cpe_activedirectory']['flag_to_profile_map']
+    profile_root = root.split('_').map(&:capitalize).join(' ')
+    profile_key = flag_to_profile_map[root][key]
+    return nil if profile_key.nil?
+
+    return nil if ad_state.nil?
 
-  flags_to_run = []
+    root_hash = ad_state[profile_root]
+    return nil if root_hash.nil?
 
-  node['cpe_activedirectory']['options'].each do |root, settings|
-    next unless [root, 'all'].include? new_resource.scope
+    root_hash[profile_key] == value
+  end
 
-    settings.reject { |_k, v| v.nil? }.each do |key, value|
-      already_set = already_set?(root, key, value)
-      if already_set.nil?
-        log "cpe_activedirectory option #{key} is not supported" do
+  def bind
+    bind_options = node['cpe_activedirectory']['bind_options']
+    options_missing = false
+    bind_options.each do |option|
+      if option.nil? || option.empty?
+        log "cpe_activedirectory binding_option #{option} is missing" do
           level :warn
         end
-        next
-      elsif !already_set
-        flags_to_run << "-#{key} #{shell_format(value)}"
+        options_missing = true
+      end
+    end
+
+    if options_missing
+      return
+    end
+
+    if node['cpe_activedirectory']['bind_method'] == 'profile_resource'
+      bind_profile(bind_options)
+    else
+      bind_execute(bind_options)
+    end
+  end
+
+  def bind_execute(bind_options)
+    # Escape everything in case of special characters
+    cmd = '/usr/sbin/dsconfigad '\
+    "-add \'#{bind_options['HostName']}\' "\
+    "-username \'#{bind_options['UserName']}\' "\
+    "-password \'#{bind_options['Password']}\' "\
```

Comment:
I don't think I'd be comfortable enough to put these LDAP credentials in my chef repository, to be honest, but I don't see another way of working this. Which brings me to...

Reply:
Yeah it depends on how locked down the account is.